src/streamer.ts

```diff
@@ -14,18 +14,38 @@ export async function startStream(
   let queue: Promise<void>[] = [];
 
   while (true) {
-    const blockHeights = await listBlocks(s3Client, config.s3BucketName, startFromBlockHeight);
+    let blockHeights;
+    try {
+      blockHeights = await listBlocks(
+        s3Client,
+        config.s3BucketName,
+        startFromBlockHeight
+      );
+    } catch (err) {
+      console.error("Failed to list blocks. Retrying.", err);
+      continue;
+    }
     if (blockHeights.length > 0) {
       startFromBlockHeight = blockHeights[blockHeights.length - 1];
     } else {
       await sleep(2000);
       continue;
     }
     for (let blockHeight of blockHeights) {
-      const streamerMessage = await fetchStreamerMessage(s3Client, config.s3BucketName, blockHeight);
-      if (lastProcessedBlockHash && lastProcessedBlockHash !== streamerMessage.block.header.prevHash) {
-        startFromBlockHeight = blockHeight - 1;
-        continue;
+      const streamerMessage = await fetchStreamerMessage(
+        s3Client,
+        config.s3BucketName,
+        blockHeight
+      );
+      // check if we have `lastProcessedBlockHash` (might be not set only on start)
+      // compare lastProcessedBlockHash` with `streamerMessage.block.header.prevHash` of the current
+      // block (ensure we don't miss anything from S3)
+      // retrieve the data from S3 if hashes don't match and repeat the main loop step
+      if (
+        lastProcessedBlockHash &&
+        lastProcessedBlockHash !== streamerMessage.block.header.prevHash
+      ) {
+        break;
```

Comment:
https://github.com/near/near-lake-framework/blob/1eedf2d7254004d9fc74f96b0e3f98c839fe992a/src/lib.rs#L271-L276

Let's have the log message to ease the debugging, and sleep to avoid draining S3 credits.